test/functional/p2p_blocksonly.py

```diff
@@ -57,6 +56,26 @@ def run_test(self):
             self.nodes[0].p2p.wait_for_tx(txid)
             assert_equal(self.nodes[0].getmempoolinfo()['size'], 1)
 
+        self.log.info('Check that txs from whitelisted peers are not rejected and relayed to others')
+        self.log.info("Restarting node 0 with whitelist permission and blocksonly")
+        self.restart_node(0, ["-persistmempool=0", "-whitelist=127.0.0.1", "-whitelistforcerelay", "-blocksonly"])
+        assert_equal(self.nodes[0].getrawmempool(),[])
+        first_peer = self.nodes[0].add_p2p_connection(P2PInterface())
+        second_peer = self.nodes[0].add_p2p_connection(P2PInterface())
+        peerInfo = self.nodes[0].getpeerinfo()[0]
+        assert_equal(peerInfo['whitelisted'], True)
+        assert_equal(peerInfo['permissions'], ['noban', 'forcerelay', 'relay', 'mempool'])
+        assert_equal(self.nodes[0].testmempoolaccept([sigtx])[0]['allowed'], True) 
+        txid = self.nodes[0].testmempoolaccept([sigtx])[0]['txid']
+        first_peer.send_message(msg_tx(FromHex(CTransaction(), sigtx)))
+
+        self.log.info('Check that the whitelisted peer is still connected after sending the transaction')
+        assert_equal(first_peer.is_connected, True)
+       
+        self.log.info('Check that the tx from whitelisted first_peer is relayed to others (ie.second_peer)')
+        second_peer.wait_for_tx(txid)
+        assert_equal(self.nodes[0].getmempoolinfo()['size'], 1)
+        self.log.info("Whitelisted peer's transaction is accepted and relayed")
```

Comment:
I noticed a pattern of using "with" statement, or context manager, for checks like this. For example, line (42 and 54). But I am not quite sure whether that is necessary. wait_for_tx will timeout after 60 seconds anyways. What resources are we trying to release with the "with" anyways?

Reply:
It is optional. If you want you can check that an exact code path was hit that also contained a debug statement.

In this case it would be:

```py
with self.nodes[0].assert_debug_log(["Force relaying tx {} from whitelisted peer=0".format(txid)]):
  first_peer.send(...
  ...
```

To check you are hitting the right branch in `src/net_processing.cpp`, where it say "Force relaying tx %s ..."